tools/src/main/java/org/apache/kafka/tools/EndToEndLatency.java

```diff
@@ -0,0 +1,226 @@
+        if (records.isEmpty()) {
+            consumer.commitSync();
+            throw new RuntimeException("poll() timed out before finding a result (timeout:[" + POLL_TIMEOUT_MS + "])");
+        }
+
+        //Check result matches the original record
+        String sent = new String(message, StandardCharsets.UTF_8);
+        String read = new String(records.iterator().next().value(), StandardCharsets.UTF_8);
+
+        if (!read.equals(sent)) {
+            consumer.commitSync();
+            throw new RuntimeException("The message read [" + read + "] did not match the message sent [" + sent + "]");
+        }
+
+        //Check we only got the one message
+        if (records.count() != 1) {
+            int count = records.count();
+            consumer.commitSync();
+            throw new RuntimeException("Only one result was expected during this test. We found [" + count + "]");
+        }
+    }
+
+    private static void setupConsumer(String topic, KafkaConsumer<byte[], byte[]> consumer) {
+        List<TopicPartition> topicPartitions = consumer.
+                partitionsFor(topic).
+                stream().map(p -> new TopicPartition(p.topic(), p.partition()))
+                .collect(Collectors.toList());
+        consumer.assign(topicPartitions);
+        consumer.seekToEnd(topicPartitions);
+        consumer.assignment().forEach(consumer::position);
+    }
+
+    private static void printResults(int numMessages, double totalTime, long[] latencies) {
+        System.out.printf("Avg latency: %.4f ms%n", totalTime / numMessages / 1000.0 / 1000.0);
+        Arrays.sort(latencies);
+        int p50 = (int) latencies[(int) (latencies.length * 0.5)];
+        int p99 = (int) latencies[(int) (latencies.length * 0.99)];
+        int p999 = (int) latencies[(int) (latencies.length * 0.999)];
+        System.out.printf("Percentiles: 50th = %d, 99th = %d, 99.9th = %d%n", p50, p99, p999);
+    }
+
+    private static byte[] randomBytesOfLen(Random random, int length) {
+        byte[] randomBytes = new byte[length];
+        Arrays.fill(randomBytes, Integer.valueOf(random.nextInt(26) + 65).byteValue());
+        return randomBytes;
+    }
+
+    private static void createTopic(Optional<String> propertiesFile, String brokers, String topic) throws IOException {
+        short defaultReplicationFactor = 1;
+        int defaultNumPartitions = 1;
+
+        System.out.printf("Topic \"%s\" does not exist. " +
+                        "Will create topic with %d partition(s) and replication factor = %d%n",
+                topic, defaultNumPartitions, defaultReplicationFactor);
+
+        Properties adminProps = loadPropsWithBootstrapServers(propertiesFile, brokers);
+        Admin adminClient = Admin.create(adminProps);
+        NewTopic newTopic = new NewTopic(topic, defaultNumPartitions, defaultReplicationFactor);
+        adminClient.createTopics(Collections.singletonList(newTopic));
```

Comment:
You need to remove this line to avoid `TopicExistsException` when the test topic does not exists. Once you do that, then I'll approve.

Reply:
Sorry that was a miss. Removed it.